Add an update-status command.

anemirovsky opened the pull request:
Fixes #350. This command is similar to the `update` command but only lists the available updates. It mirrors the `drush pm-updatestatus` command. It allows for outputing a flat list of module names to feed into another process, like an ansible script. It also includes an option to list only updates that are flagged as security releases.

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -149,38 +213,35 @@ function update_bee_callback($arguments) {
-      $folder = backdrop_get_path($item['project_type'], $item['name']);
-      if (is_dir($folder)) {
-         bee_delete($folder);
+      if ((isset($options['security-only']) || isset($options['s']) && $item['status'] !== UPDATE_NOT_SECURE) {
```
I think this is the cause of the PHP syntax error.

```suggestion
      if (isset($options['security-only']) || isset($options['s']) && $item['status'] !== UPDATE_NOT_SECURE) {
```

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -123,9 +149,47 @@ function update_db_bee_callback() {
+  }
+}
+
+function update_bee_gather_available_updates($arguments, $options) {
```
Need a function doc for all functions including helper functions.  For helper functions we tend to use the verb 'get' rather than 'gather', so for consistency, please change to that.

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -149,38 +213,35 @@ function update_bee_callback($arguments) {
+  return $data;
+}
+
+function update_bee_render_table_output($data) {
```
function doc needed here please.

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -149,38 +213,35 @@ function update_bee_callback($arguments) {
+    ),
+  ));
+}
+
```
function doc needed here please.

yorkshire-pudding reviewed (changes requested):
Hi @anemirovsky - I've added a few preliminary comments to get started, but the PR needs rebasing before I'll be able to test this as it is showing merge conflicts.

anemirovsky committed:
Add an update-status command.

Fixes #350. This command is similar to the `update` command but only
lists the available updates. It mirrors the `drush pm-updatestatus`
command. It allows for outputing a flat list of module names to feed
into another process, like an ansible script. It also includes an option
to list only updates that are flagged as security releases.

Conflicts:
  commands/update.bee.inc

anemirovsky committed:
Add a short 's' option for '--security-only'

anemirovsky committed:
Change bee 'gather' to 'get'.

anemirovsky committed:
Add documentation to functions.

anemirovsky commented:
I've rebased, corrected the syntax error and added documentation to the functions. Thank you!

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -41,6 +41,32 @@ function update_bee_command() {
+      'multiple_argument' => 'projects',
+      'options' => array(
+        'security-only' => array(
+          'description' => bt('Only update modules that have security updates available.'),
```
```suggestion
          'description' => bt('Only list updates for modules that have security updates available.'),
```

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -133,6 +159,55 @@ function update_db_bee_callback() {
+  if ($data === NULL && !$is_list_format) {
+    bee_message(bt('No Modules or Themes to Update'));
+    return;
+  }
```
If there are no updates, either because none at all or security only selected  and format is default or set to 'table' then it tries to output a table and throws several warnings (at least in does in PHP 8.2, which is the latest we currently support).

```
These are the items being updated:

 [!] Warning: Undefined variable $rows
in update_bee_render_table_output() (line 260 of /app/commands/update.bee.inc).

 [!] Warning: foreach() argument must be of type array|object, null given
in bee_get_column_widths() (line 293 of /app/includes/render.inc).

| Module | Existing | Latest |

 [!] Warning: foreach() argument must be of type array|object, null given
in bee_render_table() (line 241 of /app/includes/render.inc).
```

Also, the operator should just be two `==` as the function providing $data ( [`update_calculate_project_data()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21update%21update.compare.inc/function/update_calculate_project_data/1) ) returns an array so we are comparing an empty array with NULL which is equal (`==`) but not identical (`===`), modifying this makes it return the message.

```suggestion
  if ($data == NULL) {
    bee_message(bt('No Modules or Themes to Update'));
    return;
  }
```
If however, you want to return nothing, not even a message, when format is selected then could add a separate check for `$is_list_format` before displaying the message, but we need to return before trying to render an empty list.

```suggestion
  if ($data == NULL) {
    if (!$is_list_format) {
      bee_message(bt('No Modules or Themes to Update'));
    }
    return;
  }
```

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -41,6 +41,32 @@ function update_bee_command() {
         'bee update webform tatsu' => bt('Updates the Webform module and Tatsu theme only'),
       ),
     ),
+    'update-status' => array(
```
For consistency the command callback should match the main command name. I personally feel that update list is more descriptive of what it is; status might make me think it is just a True or False, or some categorised status. Happy to keep the others for consistency with drush.

```suggestion
    'update-list' => array(
```

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -41,6 +41,32 @@ function update_bee_command() {
+    'update-status' => array(
+      'description' => bt('Lists available updates for backdrop, modules, themes and layouts with new releases'),
+      'callback' => 'update_list_bee_callback',
+      'aliases' => array('ups', 'pm-updatestatus'),
```
```suggestion
      'aliases' => array('ups', 'pm-updatestatus', 'update-status', 'upl'),
```

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -41,6 +41,32 @@ function update_bee_command() {
+      'examples' => array(
+        'bee update-status' => bt('List updates for everything with a new release.'),
+        'bee update-status --security-only' => bt('List only security updates for everything with a new release.'),
+        'bee update-status webform tatsu' => bt('List updates for the Webform module and Tatsu theme only'),
```
```suggestion
        'bee update-list' => bt('List updates for everything with a new release.'),
        'bee update-list --security-only' => bt('List only security updates for everything with a new release.'),
        'bee update-list webform tatsu' => bt('List updates for the Webform module and Tatsu theme only'),
```

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -161,36 +236,46 @@ function update_bee_callback($arguments, $options) {
+ * @param array $data
+ *   An array of projects with updates available.
+ *
+ * @return void
```
According to our [docs](https://docs.backdropcms.org/doc-standards#functions), "If there is no return value for a function, there must not be a `@return` tag."

```suggestion
```

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -161,36 +236,46 @@ function update_bee_callback($arguments, $options) {
+ * @param array $data
+ *   An array of projects with updates available.
+ *
+ * @return void
```
According to our [docs](https://docs.backdropcms.org/doc-standards#functions), "If there is no return value for a function, there must not be a `@return` tag."

```suggestion
```

yorkshire-pudding reviewed (changes requested):
Now I've had chance to test, there are a few more actions for you.
I think this will be a useful addition to the overall feature. Thanks.

yorkshire-pudding reviewed (changes requested):
I can't suggest a change here, but this line
234
```php
if ((isset($options['security-only']) || isset($options['s'])) && $item['status'] !== UPDATE_NOT_SECURE) {
```

can now be:

```php
if (isset($options['security-only'] && $item['status'] !== UPDATE_NOT_SECURE) {
```

in #357, I improved the command handling to convert short form options to their long form equivalents before they are passed to the command callback.

anemirovsky committed:
Updates related to PR code review recommendations.

anemirovsky commented:
I've made these changes. Thanks for the notes! And the change you made in #357 is great. I was thinking of working on that a bit down the road.

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -133,6 +159,57 @@ function update_db_bee_callback() {
+ * @return array|NULL
+ *   An array of projects with updates available, or NULL if there are none.
+ */
+function update_bee_get_available_updates($arguments, $options) {
```
```suggestion
function update_bee_get_available_updates(array $arguments, array $options) {
```

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -156,41 +233,47 @@ function update_bee_callback($arguments, $options) {
+ * @param array $data
+ *   An array of projects with updates available.
+ */
+function update_bee_render_table_output($data) {
```
```suggestion
function update_bee_render_table_output(array $data) {
```

yorkshire-pudding commented on commands/update.bee.inc:
```diff
@@ -156,41 +233,47 @@ function update_bee_callback($arguments, $options) {
+ * @param array $data
+ *   An array of projects with updates available.
+ */
+function update_bee_render_list_output($data) {
```
```suggestion
function update_bee_render_list_output(array $data) {
```

yorkshire-pudding reviewed (changes requested):
@anemirovsky - you'll notice a few extra PHPCS warnings now appearing; this is due to #356 where I decided to include the PHPCS warnings as well.

These are quite straightforward; just need a type hint in the parameter declaration, in this case they are all arrays.

anemirovsky committed:
Add type hinting to function arguments per PHPCS warnings.

anemirovsky commented:
I think I've got them added now but let me know if there are any others missing. Thank you.

yorkshire-pudding reviewed (approved):
Thanks @anemirovsky - all looking good now.
A great addition to this feature.
